ctapipe/reco/hillas.py

```diff
@@ -181,73 +306,595 @@ def hillas_parameters_2(pix_x, pix_y, image):
+    #                         pix_y,
+    #                         pix_x * pix_x,
+    #                         pix_y * pix_y,
+    #                         pix_x * pix_y]) * image
+    momdata = static_pix.pixdata * image
     moms = momdata.sum(axis=1) / size
 
+    if higherMoments:
+        momdataHO = static_pix.pixdataHO * image
+        momsHO = momdataHO.sum(axis=1) / size
+
+    # give the moms values comprehensible names
+    xm, ym, x2m, xym, y2m = moms
+    if higherMoments:
+        x3m, x2ym, xy2m, y3m, x4m, x3ym, x2y2m, xy3m, y4m = momsHO
+
+    # intermediate variables (could be avoided if compiler which understands powers, etc)
+    xm2 = xm * xm
+    ym2 = ym * ym
+    xmym = xm * ym
+
     # calculate variances
 
-    vx2 = moms[2] - moms[0] ** 2
-    vy2 = moms[3] - moms[1] ** 2
-    vxy = moms[4] - moms[0] * moms[1]
+    vx2 = x2m - xm2
+    vy2 = y2m - ym2
+    vxy = xym - xmym
+
+    if higherMoments:
+        vx3 = x3m - 3.0 * xm * x2m + 2.0 * xm2 * xm
+        vx2y = x2ym - x2m * ym - 2.0 * xym * xm + 2.0 * xm2 * ym
+        vxy2 = xy2m - y2m * xm - 2.0 * xym * ym + 2.0 * xm * ym2
+        vy3 = y3m - 3.0 * ym * y2m + 2.0 * ym2 * ym
+
+    # polar coordinates of centroid
+
+    rr = np.sqrt(xm2 + ym2)  # could use hypot(xm, ym), but already have squares
+    phi = np.arctan2(ym, xm)
 
     # common factors:
 
     dd = vy2 - vx2
-    zz = np.sqrt(dd ** 2 + 4.0 * vxy ** 2)
+    zz = np.hypot(dd, 2.0 * vxy) # for simpler formulae for length & width suggested CA 901019
+
+    # shower shape parameters
 
-    # miss
+    length = np.sqrt((vx2 + vy2 + zz)/ 2.0)
+    width = np.sqrt((vx2 + vy2 - zz)/2.0)
+    # azwidth = np.sqrt(x2m + y2m - zz) #Hillas Azwidth not used anymore
+    # d = y2m - x2m
+    # z = np.sqrt(d * d + 4 * xym * xym)
+    # akwidth = np.sqrt((x2m + y2m - z) / 2.0) # Akerlof azwidth (910112) not used anymore either
+
+    # miss, simpler formula for miss introduced CA, 901101; revised MP 910112
 
     uu = 1.0 + dd / zz
     vv = 2.0 - uu
-    miss = np.sqrt((uu * moms[0] ** 2 + vv * moms[1] ** 2) / 2.0
-                   - moms[0] * moms[1] * 2.0 * vxy / zz)
+    miss = np.sqrt((uu * xm2 + vv * ym2) / 2.0
+                   - xmym * 2.0 * vxy / zz)
+    # maybe for case zz = 0, put miss = dist?
 
-    # shower shape parameters
+    # rotation angle of ellipse relative to centroid
 
-    width = np.sqrt((vx2 + vy2 - zz)/2.0) 
-    length = np.sqrt((vx2 + vy2 + zz)/ 2.0)
-    azwidth = np.sqrt(moms[2] + moms[3] - zz)
+    tanpsi_numer = (dd + zz) * ym + (2.0 * vxy * xm)
+    tanpsi_denom = (2.0 * vxy * ym) - (dd - zz) * xm
+
+    psi = (np.arctan2(tanpsi_numer, tanpsi_denom)) * u.rad
+
+    if not higherMoments:
+        return MomentParameters(size=size, cen_x=xm*unit, cen_y=ym*unit,
+                            length=length*unit, width=width, r=rr*unit, phi=phi, psi=psi.to(u.deg), miss=miss*unit)
+
+    # -- Asymmetry and other higher moments
+    if length != 0.0:
```

Comment:
Maybe, but really the question is to avoid a division by exactly zero, which might give a NaN.
If the pixel values are such that the denomator value is very small, I don't think it will be a problem (since the numerator value will be small too).
But, if it's a generally accepted way to go, sure.